utils/relay-manager.js

```diff
@@ -61,11 +91,123 @@ class RelayManager {
       _gasLimit,
       _refundToken,
       _refundAddress,
-      { gas: _gasLimitRelay, gasPrice: _gasPrice, from: relayerAccount },
+      { gas, gasPrice: _gasPrice, from: relayerAccount },
     );
 
+    console.log("_gasLimit", _gasLimit);
+    console.log("gas sent", gas);
+    console.log("gas used", tx.receipt.gasUsed);
+    console.log("ratio", _gasLimit / tx.receipt.gasUsed);
+
     return tx.receipt;
   }
+
+  /* Returns the gas limit to use as gasLimit parameter in execute function
+    + 5200  (isFeatureAuthorisedInVersionManager check)
+    + 0 / 1000 / 4800 based on which contract implements getRequiredSignatures()
+    + 2052 (getSignHash call)
+    + 45144 (checkAndUpdateUniqueness call)
+    + 10000 * number of signatures (validateSignatures call, should best be estimated but this is also close enough)
+    + Function call estimate
+    + 40000 / 30000 refund cost for 1 signatures and >1 signatures respectively
+
+    Ignoring multiplication and comparisson as that is <10 gas per operation
+  */
+  async getGasLimitRefund(_module, _method, _params, _wallet, _signers, _gasPrice) {
+    let requiredSigsGas = 0;
+    const { contractName } = _module.constructor;
+    if (contractName === "ApprovedTransfer" || contractName === "RecoveryManager") {
+      requiredSigsGas = 4800;
+    } else if (contractName === "GuardianManager") {
+      requiredSigsGas = 1000;
+    }
+
+    // Estimate cost of checkAndUpdateUniqueness call
+    let nonceCheckGas = 0;
+    if (_signers.length === 1) {
+      nonceCheckGas = 6200;
+      // Most calls here are the first for wallet so default to 20K gas extra for a storage slot update
+      // We could check if the wallet nonce is empty before we add this
+      nonceCheckGas += 20000;
```

Comment:
We should add 15k and not 20k because 5k is already accounted for in the `6200` .